Code/Tools/AssetProcessor/native/FileWatcher/FileWatcher.cpp

```diff
@@ -6,154 +6,119 @@
+
+        if ((charAtB == slash1) || (charAtB == slash2))
+        {
+            if ((charAtA != slash1) && (charAtA != slash2))
+            {
+                return false;
+            }
+            ++posA;
+        }
+        else
+        {
+            if (charAtA.toLower() != charAtB.toLower())
+            {
+                return false;
+            }
+            ++posA;
+        }
+    }
+    return true;
 }
 
 //////////////////////////////////////////////////////////////////////////
 /// FileWatcher
 FileWatcher::FileWatcher()
-    : m_nextHandle(0)
+    : m_platformImpl(AZStd::make_unique<PlatformImplementation>())
 {
-    qRegisterMetaType<FileChangeInfo>("FileChangeInfo");
+    auto makeFilter = [this](auto signal)
+    {
+        return [this, signal](QString path)
+        {
+            const auto foundWatchRoot = AZStd::find_if(begin(m_folderWatchRoots), end(m_folderWatchRoots), [path](const WatchRoot& watchRoot)
+            {
+                return Filter(path, watchRoot);
+            });
+            if (foundWatchRoot == end(m_folderWatchRoots))
+            {
+                return;
+            }
+            AZStd::invoke(signal, this, path);
+        };
+    };
+
+    connect(this, &FileWatcher::rawFileAdded, makeFilter(&FileWatcher::fileAdded));
+    connect(this, &FileWatcher::rawFileRemoved, makeFilter(&FileWatcher::fileRemoved));
+    connect(this, &FileWatcher::rawFileModified, makeFilter(&FileWatcher::fileModified));
 }
 
 FileWatcher::~FileWatcher()
 {
+    disconnect();
+    StopWatching();
 }
 
-int FileWatcher::AddFolderWatch(FolderWatchBase* pFolderWatch)
+void FileWatcher::AddFolderWatch(QString directory, bool recursive)
 {
-    if (!pFolderWatch)
-    {
-        return -1;
-    }
-
-    FolderRootWatch* pFolderRootWatch = nullptr;
-
-    //see if this a sub folder of an already watched root
-    for (auto rootsIter = m_folderWatchRoots.begin(); !pFolderRootWatch && rootsIter != m_folderWatchRoots.end(); ++rootsIter)
+    // Search for an already monitored root that is a parent of `directory`,
+    // that is already watching subdirectories recursively
+    const auto found = AZStd::find_if(begin(m_folderWatchRoots), end(m_folderWatchRoots), [directory](const WatchRoot& root)
     {
-        if (FolderWatchBase::IsSubfolder(pFolderWatch->m_folder, (*rootsIter)->m_root))
-        {
-            pFolderRootWatch = *rootsIter;
-        }
-    }
+        return root.m_recursive && IsSubfolder(directory, root.m_directory);
+    });
 
-    bool bCreatedNewRoot = false;
-    //if its not a sub folder
-    if (!pFolderRootWatch)
+    if (found != end(m_folderWatchRoots))
     {
-        //create a new root and start listening for changes
-        pFolderRootWatch = new FolderRootWatch(pFolderWatch->m_folder);
-
-        //make sure the folder watcher(s) get deleted before this
-        pFolderRootWatch->setParent(this);
-        bCreatedNewRoot = true;
+        // This directory is already watched
```

Comment:
I might be misunderstanding the code here, but how does it work if a parent watch is removed later?

Reply:
`FileWatcher` doesn't support removing directories from the watch, only clearing *all* the watches.